src/sentry/api/bases/dashboard.py

```diff
@@ -0,0 +1,98 @@
+from __future__ import absolute_import
+from rest_framework import serializers
+from sentry.models import Dashboard, Widget, WidgetDisplayTypes, WidgetDataSourceTypes
+
+from sentry.api.serializers.rest_framework.json import JSONField
+from sentry.api.serializers.rest_framework.list import ListField
+from sentry.api.fields.user import UserField
+from sentry.api.bases.discoversavedquery import DiscoverSavedQuerySerializer
+
+
+class WidgetDataSource(serializers.WritableField):
+    name = serializers.CharField(required=False)
+    data = JSONField(required=True)
+    type = serializers.CharField(required=True)
+
+    def validate_type(self, attrs, source):
+        type = attrs[source]
+        if type not in WidgetDataSourceTypes.__members__:
+            raise ValueError('Widget data source type %s not recognized.' % type)
+        return attrs
+
+    def validate(self, data):
+        if data['type'] == WidgetDataSourceTypes.DISCOVER_SAVED_SEARCH:
+            serializer = DiscoverSavedQuerySerializer(data=data['data'])
+            if not serializer.is_valid():
+                raise ValueError('Error validating DiscoverSavedQuery: %s' % serializer.errors)
+        return data
+
+
+class WidgetSerializer(serializers.WritableField):
+    order = serializers.IntegerField(min_value=0, required=True)
+    display_type = serializers.CharField(required=True)
+    data = JSONField(
+        required=True,
+    )
+    data_sources = ListField(
+        child=WidgetDataSource(),
+        required=True,
+        default=[],
+    )
+    data = JSONField(
+        required=False,
+    )
+
+    def validate_display_type(self, attrs, source):
+        display_type = attrs[source]
+        if display_type not in WidgetDisplayTypes.__members__:
+            raise ValueError('Widget display_type %s not recognized.' % display_type)
+        return attrs
+
+
+class DashboardSerializer(serializers.Serializer):
+    title = serializers.CharField(required=True)
+    owner = UserField(required=True)
+    data = JSONField(
+        required=False,
+    )
+    widgets = ListField(
+        child=WidgetSerializer(),
+        required=False,
+        default=[],
+    )
+
+    def save(self):
+        dashboard_id = self.context.get('dashboard_id')
```

Comment:
This is probably completely wrong, but I was just trying to get something working